### PR TITLE
Cleanup how text providers are set up and synced

### PR DESCRIPTION
**Pull request by Jacalz**
### Description:

This cleans up a lot of code for setting up and syncing changes between widgets and their text providers. Text providers are created inline in the struct to not require an extra allocation plus nil checks, segments are synced without needing to allocate garbage in the form of new slices and some other cleanups are included. 

### Checklist:

- [ ] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

### CODE REVIEW

File: widget/entry.go [hunk 1]
```diff
@@ -1009,16 +1009,17 @@ func (e *Entry) getRowCol(p fyne.Position) (int, int) {
...
+	text := e.textProvider()
+	rowHeight := text.charMinSize(e.Password, e.TextStyle).Height
 	row := int(math.Floor(float64(p.Y+e.scroll.Offset.Y-theme.LineSpacing()) / float64(rowHeight)))
 	col := 0
 	if row < 0 {
 		row = 0
-	} else if row >= e.textProvider().rows() {
-		row = e.textProvider().rows() - 1
-		col = e.textProvider().rowLength(row)
+	} else if row >= text.rows() {
```
Comment (dweymouth) [inline widget/entry.go:1018]: Should we call text.rows() just once?

Comment (Jacalz) [reply widget/entry.go:1019]: I checked and doesn't matter. It us just returning the length of a slice and that method will be inlined.

**Comment by Jacalz** [comment]
Rebased on latest develop now.

**Comment by Jacalz** [comment]
Sorry for the force push. Did a merge first and then another because it still didn't think that the conflicts were resolved. After the second merge, it still didn't believe that I fixed them so I gave up and rebased instead.

**Comment by Jacalz** [comment]
I don't understand why the tests are failing like they are. The code worked fine before the merge/rebase.

**Comment by andydotxyz** [comment]
> I don't understand why the tests are failing like they are. The code worked fine before the merge/rebase.

No idea. the base branches are all working correctly too.

**Comment by Jacalz** [comment]
Will have a look again today or tomorrow. A fresh mind might give some insights

**Comment by Jacalz** [comment]
Closing in favour of #4649

**Pull request closed without merging**